### PR TITLE
Restyle file uploads to fit in with modern forms UI

### PR DESCRIPTION
**Pull request by janfaracik**
A small PR to update the file upload control so that it fits in with the rest of forms. The button now resembles a standard .jenkins-button rather than the native OS one.

**Before**

image

**After**

image

### Testing done

Go to a page with a file upload control, such as Plugin Manager -> Advanced.
Spot the file upload control.
Upload a file.
File uploads successfully.

### Proposed changelog entries

- Restyle file uploads to fit in with modern forms UI

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)).
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@jenkinsci/sig-ux 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7452"></a>

### CODE REVIEW

**Review by NotMyFault — COMMENTED** [review]
Screenshot 2022-11-26 at 10 23 28
Screenshot 2022-11-26 at 10 23 34

Is it intended to truncate the file name? Looks a bit odd to me. So is the missing space between the button and the file to upload.

Maybe we should add a page to the design lib outlining that you should use the `jenkins-file-upload` class when accepting file inputs, rather than jenkins-upload or setting-upload 🤔

**Comment by janfaracik** [comment]
> Screenshot 2022-11-26 at 10 23 28 Screenshot 2022-11-26 at 10 23 34
> 
> Is it intended to truncate the file name? Looks a bit odd to me

image Good spot, hadn't tried it with a really long file name - just needed to set the width to 100%.

> So is the missing space between the button and the file to upload.

Yeah it's a limitation with the upload input :/ I've tried everything I can think of but still no additional padding.

> Maybe we should add a page to the design lib outlining that you should use the `jenkins-file-upload` class when accepting file inputs, rather than jenkins-upload or setting-upload 🤔

Definitely 👍

**Review by NotMyFault — APPROVED** [review]

**Comment by daniel-beck** [comment]
Would it not be easier / more backwards compatible to apply different rules for `input type="file"` rather than requiring every declaration to be adapted to what amounts to redundant information? Or would there be `input type="file"` that should not have `class="jenkins-file-upload"`?

**Comment by uhafner** [comment]
> Would it not be easier / more backwards compatible to apply different rules for `input type="file"` rather than requiring every declaration to be adapted to what amounts to redundant information? Or would there be `input type="file"` that should not have `class="jenkins-file-upload"`?

Actually this calls for a jelly template {{f:fileButton}} as well (encapsulation).

**Comment by daniel-beck** [comment]
> Actually this calls for a jelly template `f:fileButton` as well (encapsulation). 

Not opposed (I've previously impersonated a broken record asking for Jelly components instead of changing CSS classes around), but _only_ doing that still wouldn't take care of existing upload fields.

**Comment by timja** [comment]
I've filed PRs with all plugins that look to be maintained using this search: https://github.com/search?q=org%3Ajenkinsci+input+type%3D%22file%22+NOT+repo%3Ajenkinsci%2Fjenkins&type=code&p=1

**Review by NotMyFault — APPROVED** [review]
Looks good, left some optional feedback for consideration.

**Review by timja — APPROVED** [review]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

File: core/src/main/resources/lib/form/file.jelly
Comment (jglick) [inline core/src/main/resources/lib/form/file.jelly:37]: Optional I guess?

You could just copy unknown tag attrs to the HTML tag. There is some Jelly/Stapler trick to do that.

**Comment by jglick** [comment]
> I've filed PRs with all plugins that look to be maintained using this search

Looks like you missed `file-parameters-plugin`, because you did not search for an intervening attribute (`name` in that case).

File: core/src/main/resources/lib/form/file.jelly
Comment (timja) [reply core/src/main/resources/lib/form/file.jelly:37]: Yes, `use="required"` denotes mandatory.

if can you point to it that could be useful although likely unneeded

Comment (strangelookingnerd) [reply core/src/main/resources/lib/form/file.jelly:37]: > You could just copy unknown tag attrs to the HTML tag. There is some Jelly/Stapler trick to do that.

Second this as it is important for my use case.

Comment (jglick) [reply core/src/main/resources/lib/form/file.jelly:37]: https://github.com/jenkinsci/jenkins/blob/7c1885554e394176dacd6296236403a3181ef351/core/src/main/java/hudson/util/jelly/MorphTagLibrary.java#L24-L32 https://github.com/jenkinsci/jenkins/blob/7c1885554e394176dacd6296236403a3181ef351/core/src/main/resources/lib/form/textbox.jelly#L107-L113

Comment (timja) [reply core/src/main/resources/lib/form/file.jelly:37]: Thanks, @strangelookingnerd what is missing that you need?

Comment (strangelookingnerd) [reply core/src/main/resources/lib/form/file.jelly:37]: See https://github.com/jenkinsci/custom-folder-icon-plugin/blob/master/src/main/resources/jenkins/plugins/foldericon/CustomFolderIcon/config.jelly#L34

I'm adapting the size (for visual reasons) and add an ID to add a change event in the following `<script>` section 
https://github.com/jenkinsci/custom-folder-icon-plugin/blob/5f60c7af85f7dfd56caca7e6ce392e70442ab616/src/main/resources/jenkins/plugins/foldericon/CustomFolderIcon/config.jelly#L61-L69

Maybe this is the wrong approach but without the ID my use case breaks.

Comment (timja) [reply core/src/main/resources/lib/form/file.jelly:37]: Styling should be done via classes and not inline css, I added support for an ID because of your plugin but you could also do that via a CSS class anyway

**Pull request merged**
